core/FlushManager.hpp

```diff
@@ -38,7 +42,66 @@ namespace olympia
     class FlushManager : public sparta::Unit
     {
     public:
-        typedef uint64_t FlushingCriteria;
+
+        enum class FlushCause : std::uint16_t {
+            TRAP = 0,
+            __FIRST = TRAP,
+            MISPREDICTION,
+            TARGET_MISPREDICTION,
+            MISFETCH,
+            POST_SYNC,
+            UKNOWN,
+            __LAST
+        };
+
+        class FlushingCriteria
+        {
+        public:
+            FlushingCriteria(FlushCause cause, InstPtr inst_ptr) :
+                cause_(cause),
+                inst_ptr_(inst_ptr) {}
+
+            FlushingCriteria() = default;
+            FlushingCriteria(const FlushingCriteria &rhs) = default;
+            FlushingCriteria &operator=(const FlushingCriteria &rhs) = default;
+
+            FlushCause getCause() const        { return cause_; }
+            const InstPtr & getInstPtr() const { return inst_ptr_; }
+
+            bool isInclusiveFlush() const
+            {
+                static const std::map<FlushCause, bool> inclusive_flush_map = {
+                    {FlushCause::TRAP,                 true},
+                    {FlushCause::MISFETCH,             true},
+                    {FlushCause::MISPREDICTION,        false},
+                    {FlushCause::TARGET_MISPREDICTION, false},
+                    {FlushCause::POST_SYNC,            false}
+                };
```

Comment:
This is fine for now... but I wonder if this should be the decision of the instruction causing the flush...

Reply:
Did think about doing it that way, but then I thought against it as I figured it would be more robust tying the cause and inclusiveness of the flush together. I don't feel too strongly about it, but perhaps we could keep it this way for now?

Reply:
Yeah, I actually like this approach, so feel free to keep it this way for now.